rust-runtime/aws-smithy-observability/src/meter.rs

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Metrics are used to gain insight into the operational performance and health of a system in
+//! real time.
+
+use crate::attributes::{Attributes, Context};
+use crate::error::ObservabilityError;
+
+/// Provides named instances of [Meter].
+pub trait MeterProvider {
+    /// Get or create a named [Meter].
+    fn get_meter(&self, scope: &'static str, attributes: Option<&Attributes>) -> Box<dyn Meter>;
```

Comment:
**discuss** Should this be `-> impl Meter`?

Reply:
I don't think that would work. Returning `impl Meter` would make `MeterProvider` not object safe and so we couldn't pass it around as a `dyn` everywhere. 

From [docs](https://doc.rust-lang.org/reference/items/traits.html#object-safety):

- All associated functions must either be dispatchable from a trait object or be explicitly non-dispatchable:
  - Dispatchable functions must:
    - Not have an opaque return type; that is,
      - Not have a return position impl Trait type (fn example(&self) -> impl Trait).